database/src/main/kotlin/fi/vauhtijuoksu/vauhtijuoksuapi/database/impl/TimerDatabase.kt

```diff
@@ -0,0 +1,130 @@
+import io.vertx.sqlclient.templates.SqlTemplate
+import mu.KotlinLogging
+import java.text.SimpleDateFormat
+import java.time.Instant
+import java.time.format.DateTimeFormatter
+import java.util.Date
+import java.util.UUID
+import javax.inject.Inject
+
+class TimerDatabase
+@Inject constructor(
+    private val client: SqlClient,
+    configuration: DatabaseConfiguration
+) : AbstractModelDatabase<Timer, TimerDbModel>(
+    client,
+    configuration,
+    "timers",
+    null,
+    { TimerDbModel -> TimerDbModel.toTimer() }
+),
+    VauhtijuoksuDatabase<Timer> {
+    private val logger = KotlinLogging.logger {}
+    val df = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+
+    override fun <I, R> mapToFunction(template: SqlTemplate<I, R>): SqlTemplate<I, RowSet<TimerDbModel>> {
+        return template.mapTo(TimerDbModel::class.java)
+    }
+
+    @Suppress("MaxLineLength") // SQL is prettier without too many splits
+    override fun add(record: Timer): Future<Timer> {
+        return SqlTemplate.forUpdate(
+            client,
+            "INSERT INTO timers " +
+                "(start_time, end_time) VALUES " +
+                "(#{start_time}, #{end_time}) " +
+                "RETURNING *"
+        )
+            .mapFrom(TimerDbModel::class.java)
+            .mapTo(TimerDbModel::class.java)
+            .execute(TimerDbModel.fromTimer(record))
+            .recover {
+                throw ServerError("Failed to insert $record because of ${it.message}")
+            }
+            .map {
+                logger.debug { "Inserted timer $record" }
+                return@map it.iterator().next().toTimer()
+            }
+    }
+
+    @Suppress("MaxLineLength") // SQL is prettier without too many splits
+    override fun addMany(records: List<Timer>): Future<List<Timer>?> {
+        fun insertStatement(data: List<Timer>): String {
+            fun valuesStringForTimer(timer: Timer): String {
+                val startTime = if (timer.startTime != null) "'${df.format(timer.startTime)}'" else null
+                val endTime = if (timer.endTime != null) "'${df.format(timer.endTime)}'" else null
+                @Suppress("MaxLineLength")
+                return "('${timer.id}', $startTime, $endTime)"
+            }
+
+            var statement = "INSERT INTO timers VALUES "
+            for (timer in data) {
+                statement += "${valuesStringForTimer(timer)},"
+            }
+            statement = statement.trim(',')
+            statement += """ ON CONFLICT (id) DO UPDATE 
+                SET start_time = excluded.start_time, 
+                end_time = excluded.end_time;"""
+            return statement
+        }
+
+        return client.query(insertStatement(records))
+            .execute()
+            .recover {
+                throw ServerError("Failed to insert $records because of ${it.message}")
+            }
+            .map {
+                return@map it.toList().map { row -> toTimer(row) }
```

Comment:
I think you could use `toTimer` with `mapFrom` to make the code more in line with existing code